File: core/src/main/cpp/entityauth/PresharedProfileAuthenticationData.cpp
/**
 * Copyright (c) 2016-2017 Netflix, Inc.  All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <entityauth/EntityAuthenticationScheme.h>
#include <entityauth/PresharedProfileAuthenticationData.h>
#include <io/MslEncoderException.h>
#include <io/MslEncoderFactory.h>
#include <io/MslObject.h>
#include <Macros.h>
#include <MslEncodingException.h>
#include <MslError.h>
#include <util/MslUtils.h>
#include <cassert>

using namespace std;
using namespace netflix::msl;
using namespace netflix::msl::entityauth;
using namespace netflix::msl::io;
using namespace netflix::msl::util;

namespace netflix {
namespace msl {
namespace entityauth {

namespace {
/** Key entity preshared keys identity. */
const string KEY_PSKID = "pskid";
/** Key entity profile. */
const string KEY_PROFILE = "profile";

/** Identity concatenation character. */
const string CONCAT_CHAR = "-";

} // namespace anonymous


PresharedProfileAuthenticationData::PresharedProfileAuthenticationData(const string& pskid, const string& profile)
: EntityAuthenticationData(EntityAuthenticationScheme::PSK_PROFILE)
, pskid(pskid)
, profile(profile)
{
}

PresharedProfileAuthenticationData::PresharedProfileAuthenticationData(shared_ptr<MslObject> authMo)
: EntityAuthenticationData(EntityAuthenticationScheme::PSK_PROFILE)
{
    assert(authMo);
    try {
        pskid = authMo->getString(KEY_PSKID);
        profile = authMo->getString(KEY_PROFILE);
    } catch (const MslEncoderException& e) {
        throw MslEncodingException(MslError::MSL_PARSE_ERROR, "psk profile authdata " + authMo->toString(), e);
    }
}

string PresharedProfileAuthenticationData::getIdentity() const
{
    return pskid + CONCAT_CHAR + profile;
}

shared_ptr<MslObject> PresharedProfileAuthenticationData::getAuthData(shared_ptr<MslEncoderFactory> encoder,
        const MslEncoderFormat&) const
{
    shared_ptr<MslObject> mo = encoder->createObject();
    mo->put(KEY_PSKID, pskid);
    mo->put(KEY_PROFILE, profile);
    return mo;
}

bool PresharedProfileAuthenticationData::equals(shared_ptr<const EntityAuthenticationData> obj) const
{
	if (!obj) return false;
    if (this == obj.get()) return true;
    if (!instanceof<const PresharedProfileAuthenticationData>(obj)) return false;
    shared_ptr<const PresharedProfileAuthenticationData> that = dynamic_pointer_cast<const PresharedProfileAuthenticationData>(obj);
    return EntityAuthenticationData::equals(obj) && pskid == that->pskid && profile == that->profile;
}

bool operator==(const PresharedProfileAuthenticationData& a, const PresharedProfileAuthenticationData& b)
{
	shared_ptr<const PresharedProfileAuthenticationData> ap(&a, &MslUtils::nullDeleter<PresharedProfileAuthenticationData>);
	shared_ptr<const PresharedProfileAuthenticationData> bp(&b, &MslUtils::nullDeleter<PresharedProfileAuthenticationData>);
	return ap->equals(bp);
}

}}} // namespace netflix::msl:entityauth

File: core/src/main/javascript/entityauth/PresharedProfileAuthenticationData.js
/**
 * Copyright (c) 2014-2017 Netflix, Inc.  All rights reserved.
 * 
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * <p>Preshared keys profile entity authentication data.</p>
 * 
 * <p>
 * {@code {
 *   "#mandatory" : [ "pskid", "profile" ],
 *   "pskid" : "string",
 *   "profile" : "string",
 * }} where:
 * <ul>
 * <li>{@code pskid} is the entity preshared keys identity</li>
 * <li>{@code profile} is the entity profile</li>
 * </ul></p>
 * 
 * @author Wesley Miaw <wmiaw@netflix.com>
 */
(function(require, module) {
    "use strict";
    
    var EntityAuthenticationData = require('../entityauth/EntityAuthenticationData.js');
    var EntityAuthenticationScheme = require('../entityauth/EntityAuthenticationScheme.js');
    var AsyncExecutor = require('../util/AsyncExecutor.js');
    var MslEncoderException = require('../io/MslEncoderException.js');
    var MslEncodingException = require('../MslEncodingException.js');
    var MslError = require('../MslError.js');
    
    /**
     * Key entity preshared keys identity.
     * @const
     * @type {string}
     */
    var KEY_PSKID = "pskid";
    /**
     * Key entity profile.
     * @const
     * @type {string}
     */
    var KEY_PROFILE = "profile";
    
    /**
     * Identity concatenation character.
     * @const
     * @type {string}
     */
    var CONCAT_CHAR = "-";

    var PresharedProfileAuthenticationData = module.exports = EntityAuthenticationData.extend({
        /**
         * Construct a new preshared keys authentication data instance from the
         * specified entity preshared keys identity and profile.
         * 
         * @param {string} pskid the entity preshared keys identity.
         * @param {string} profile the entity profile.
         */
        init: function init(pskid, profile) {
            init.base.call(this, EntityAuthenticationScheme.PSK_PROFILE);
            
            // The properties.
            var props = {
                presharedKeysId: { value: pskid, writable: false, configurable: false },
                profile: { value: profile, writable: false, configurable: false },
            };
            Object.defineProperties(this, props);
        },
    
        /**
         * <p>Returns the entity identity. This is equal to the preshared keys
         * identity and profile strings joined with a hyphen, e.g.
         * {@code pskid-profile}.</p>
         * 
         * @return {string} the entity identity.
         */
        getIdentity: function getIdentity() {
            return this.presharedKeysId + CONCAT_CHAR + this.profile;
        },
    
        /** @inheritDoc */
        getAuthData: function getAuthData(encoder, format, callback) {
            AsyncExecutor(callback, function() {
                var mo = encoder.createObject();
                mo.put(KEY_PSKID, this.presharedKeysId);
                mo.put(KEY_PROFILE, this.profile);
                return mo;
            }, this);
        },
    
        /** @inheritDoc */
        equals: function equals(that) {
            if (this === that) return true;
            if (!(that instanceof PresharedProfileAuthenticationData)) return false;
            return (equals.base.call(this, that) && this.presharedKeysId == that.presharedKeysId && this.profile == that.profile);
        },
    });

    /**
     * Construct a new preshared keys profile authentication data instance from
     * the provided MSL object.
     * 
     * @param {MslObject} authMo the authentication data MSL object.
     * @throws MslEncodingException if there is an error parsing the entity
     *         authentication data.
     */
    var PresharedProfileAuthenticationData$parse = function PresharedProfileAuthenticationData$parse(authMo) {
        try {
            var pskid = authMo.getString(KEY_PSKID);
            var profile = authMo.getString(KEY_PROFILE);
            return new PresharedProfileAuthenticationData(pskid, profile);
        } catch (e) {
            if (e instanceof MslEncoderException)
                throw new MslEncodingException(MslError.MSL_PARSE_ERROR, "psk profile authdata " + authMo);
            throw e;
        }
    };
    
    // Exports.
    module.exports.parse = PresharedProfileAuthenticationData$parse;
})(require, (typeof module !== 'undefined') ? module : mkmodule('PresharedProfileAuthenticationData'));